nestedtensor/nested/masking.py

```diff
@@ -1,111 +1,184 @@
 
-    if mask.dim() == 1:
-        tensors = [tensor[i] if mask[i]
-                   else None for i in range(len(mask))]
-        tensors = list(filter(lambda x: x is not None, tensors))
+    if tensor.numel() == 0 and mask.numel() != 0:
+        raise RuntimeError("Data tensor can't be emtpy if a mask has values.")
+
+    if tensor.numel() != 0 and mask.numel() == 0:
+        raise RuntimeError("Mask tensor can't be emtpy if a data tensor has values.")
+
+    return nt_from_tensor_mask(tensor, mask, nested_dim, tensor.dtype)
+
+
+def nt_from_tensor_mask(tensor, mask, nested_dim, dt):
+    def _merge(tensors, nested_dim):
         if len(tensors) == 0:
-            return tensor.narrow(0, 0, 0)
-        return _merge(tensors, nested_dim)
+            return torch.tensor([], dtype=dt)
+        return torch.stack(tensors)
+
+    if nested_dim == 0:
+        if (mask.numel() == 0) or (mask.numel() == 1 and mask == True):
+            return tensor
+
+        if mask.dim() == 1:
+            tensors = [tensor[i] if mask[i] else None for i in range(len(mask))]
+            tensors = list(filter(lambda x: x is not None, tensors))
+            return _merge(tensors, nested_dim)
+
+        if mask.dim() > 1:
+            tensors = [nt_from_tensor_mask(t, m, nested_dim, dt) for (t, m) in zip(tensor, mask)]
+            if not all(t.numel() == 0 for t in tensors):
+                tensors = list(filter(lambda x: x.numel() > 0, tensors))
+            return _merge(tensors, nested_dim)
+        else:
+            return None
     else:
-        sub_nested_dim = None if nested_dim is None else nested_dim - 1
-        tensors = [nested_tensor_from_tensor_mask(
-            t, m, sub_nested_dim) for (t, m) in zip(tensor, mask)]
-        if not all(t.numel() == 0 for t in tensors):
-            tensors = list(filter(lambda x: x.numel() > 0, tensors))
-        if len(tensors) == 0:
-            return tensor.narrow(0, 0, 0)
-        return _merge(tensors, nested_dim)
+        inner_tensors = []
+        if (mask.numel() == 0) or (mask.numel() == 1 and mask == True):
+            for i in range(len(tensor)):
+                inner_tensors.append(nt_from_tensor_mask(tensor[i], mask, nested_dim - 1, dt))
+        elif (mask.numel() == 1 and mask == False):
+            inner_tensors.append(None)
+        else:
+            inner_tensors = [nt_from_tensor_mask(t, m, nested_dim - 1, dt) for (t, m) in zip(tensor, mask)]
+
+        # Filtering out None values which were ignored by mask
+        inner_tensors = list(filter(lambda x: x is not None, inner_tensors))
+        return creation.nested_tensor(inner_tensors)
+
+# Get max size per each dimension from all the passed tensors.
+def get_max_size(obj, res=[1]):
+    if isinstance(obj, list) or isinstance(obj, tuple):
+        for o in obj:
+            res = get_max_size(o, res)
+
+    if isinstance(obj, nestedtensor.nested.nested.NestedTensor):
+        tres = get_max_size(obj.unbind())
```

Comment:
Instead of calling obj.unbind() pass in obj.nested_size(), which doesn't require this expensive operation (unbind splits a NestedTensor into Tensor constituents whereas you only care about their sizes).

Reply:
nested_size() returns 'nestedtensor._C.IValueNode' which is not iterable

Reply:
IValueNode supports [unbind](https://github.com/pytorch/nestedtensor/blob/master/nestedtensor/csrc/utils/python_nested_node.h#L27) to get the entries.

Reply:
perf opt, separate PR

Reply:
or maybe not